enhancements/ingress/haproxy-router-compression.md

```diff
@@ -0,0 +1,363 @@
+header to prevent backend servers from compressing responses.  This forces all response compression to happen within
+HAProxy, jeopardizing its performance.  This configuration option is not recommended, even by HAProxy guides, so it
+is omitted from this proposal.
+
+* HAProxy provides a performance tuning variable, `maxcomprate`, to set the maximum per-process input compression rate
+  to the specified number of kilobytes per second.  It works in conjunction with `tune.comp.maxlevel` to rate-limit the
+  compression level to save data bandwidth. Another, `maxcompcpuusage`, does the same to limit CPU used on compression.
+Neither variable is currently exposed, and both are omitted from this proposal.  (Other per-session performance tuning
+variables like `maxconnrate`, which set the maximum per-process number of connections per second, and `maxsslrate`,
+which sets the maximum per-process number of SSL sessions per second, are not exposed either.)
+
+## Proposal
+
+Currently, OpenShift router's `haproxy-config.template` file contains the lines below in the `defaults` section, but the mechanism for setting the
+environment variables is not exposed.
+```go
+defaults
+...
+{{- if isTrue (env "ROUTER_ENABLE_COMPRESSION") }}
+  compression algo gzip
+  compression type {{ env "ROUTER_COMPRESSION_MIME" "text/html text/plain text/css" }}
+  {{- end }}
+```
+
+### Global HTTP Compression via IngressControllerSpec and environment variables
+
+The proposed change to IngressControllerSpec exposes the HTTP compression configuration variables at the default level:
+```go
+type IngressControllerSpec struct {
+...
+    // httpCompression defines policy for HTTP traffic compression
+    // 
+    // The default value is no compression
+    //
+    // +optional
+    HTTPCompression HTTPCompressionPolicy `json:"httpCompression,omitempty"`
+}
+...
+// httpCompressionPolicy turns on compression for the specified MIME types
+//
+// This field is optional, and its absence implies that compression should not be enabled
+// globally in HAProxy.
+//
+// If httpCompressionPolicy exists, compression should be enabled only for the specified
+// MIME types
+type HTTPCompressionPolicy struct {
+	// compressionMIMETypes is a list of MIME types that should have compression applied.
+	// At least one MIME type must be specified.
+	//
+	// Note: Not all MIME types benefit from compression, but HAProxy will still use resources
+	// to try to compress if instructed to.  Generally speaking, text (html, css, js, etc.)
+	// formats benefit from compression, but formats that are already compressed (pdf, image,
+	// audio, video, etc.) benefit little in exchange for the time and cpu spent on compressing
+	// again. See https://joehonton.medium.com/the-gzip-penalty-d31bd697f1a2
+	//
+	// +required
+	// +kubebuilder:validation:MinItems=0
+	CompressionMIMETypes []CompressionMIMEType `json:"compressionMIMETypes,omitempty"`
```

Comment:
Maybe we could delete "compression" in the field name to avoid stutter (`spec.httpCompression.compressionMIMETypes` → `spec.httpCompression.mimeTypes`):
```suggestion
	CompressionMIMETypes []CompressionMIMEType `json:"mimeTypes,omitempty"`
```